### PR TITLE
dispatcher: make `approximateMonotonicTime()` more accurate

### PR DESCRIPTION
**Pull request by danzh2010**
Commit Message: update the `approximate_monotonic_time_` after polling instead of before polling. The Envoy FileEventImpl doesn't timeout the polling when there is no I/O. Updating  `approximate_monotonic_time_` before polling could cause it to be lagged behind for a substantial period. (`approximateMonotonicTime()` is only used by `EnvoyQuicClock`.) 

This is done by registering a new `LibeventScheduler` callback with `evwatch_check_new()` to do the update right before LibeventScheduler::onCheckForStats().

Additional Description: fix existing potential TSAN failure in a few integration tests.
Risk Level: low, it should at most made no difference if not improving QUIC performance.
Testing: new unit test
Docs Changes: N/A
Release Notes: Yes
Platform Specific Features: N/A
Runtime guard: envoy.restart_features.fix_dispatcher_approximate_now

### CODE REVIEW

**Review by RyanTheOptimist — DISMISSED** [review]
Thanks for fixing this!

**Comment by RyanTheOptimist** [comment]
Looks like a real TSAN failure:

https://dev.azure.com/cncf/envoy/_build/results?buildId=178370&view=logs&jobId=d1f76054-8f79-554b-6f4a-11d6a63b8e00&j=d1f76054-8f79-554b-6f4a-11d6a63b8e00&t=33c64725-b304-5065-9740-0ce244fa1a25

**Comment by danzh2010** [comment]
> Looks like a real TSAN failure:
> 
> https://dev.azure.com/cncf/envoy/_build/results?buildId=178370&view=logs&jobId=d1f76054-8f79-554b-6f4a-11d6a63b8e00&j=d1f76054-8f79-554b-6f4a-11d6a63b8e00&t=33c64725-b304-5065-9740-0ce244fa1a25

Fixed

**Comment by danzh2010** [comment]
These TSAN failure has nothing to do with dispatcher: https://dev.azure.com/cncf/envoy/_build/results?buildId=178449&view=logs&j=d1f76054-8f79-554b-6f4a-11d6a63b8e00&t=33c64725-b304-5065-9740-0ce244fa1a25&l=23569. But they are only reproducible with this PR. I'll take further look.

File: test/extensions/internal_redirect/redirect_extension_integration_test.cc [hunk 1]
```diff
@@ -16,6 +16,17 @@ constexpr char kTestHeaderKey[] = "test-header";
...
+  void TearDown() override {
+    for (auto& fake_upstream_connection : upstream_connections_) {
+      AssertionResult result = fake_upstream_connection->close();
+      RELEASE_ASSERT(result, result.message());
+      result = fake_upstream_connection->waitForDisconnect();
+      RELEASE_ASSERT(result, result.message());
+      fake_upstream_connection.reset();
+    }
+    cleanupUpstreamAndDownstream();
+  }
```
Comment (RyanTheOptimist) [inline test/extensions/internal_redirect/redirect_extension_integration_test.cc:28]: Out of curiosity, why are these required? Does this current PR tickle this behavior somehow?

**Review by RyanTheOptimist — APPROVED** [review]
